### PR TITLE
feat: replace Node.js REPL with plain vm context for script usage MONGOSH-1720

### PR DESCRIPTION
**Pull request by addaleax**
[Definitely recommend viewing the diff in whitespace-changes-disabled mode :)]

- Allow running `MongoshNodeRepl` instances either based on:
  - The existing mechanism for running code, which is spinning up
    a Node.js REPL and using it to evaluate code; or
  - A more lightweight mechanism that only creates a `vm` context
    and then run code using that context directly.
- Introduce a new command-line switch, `--jsContext`, that can be
  used to explicitly select the desired behavior, with possible values
  of `repl`, `plain-vm` and `auto`. The default behavior is to switch
  depending on whether the CLI will enter interactive mode or not.

Running in `plain-vm` mode will significantly improve runtime
performance (a 6× reduction of script run time in local testing),
coming from the removal of async context tracking that the REPL
uses to identify async operations which were originally spawned
from the REPL instance in question.

This lack of async context tracking comes with some implications,
in particular asynchronously thrown errors (in the Node.js sense,
e.g. `setImmediate(() => { throw ... })`) will lead to slightly
different behavior (e.g. different error code and error output).
That is probably acceptable breakage in this context.

### CODE REVIEW

File: packages/cli-repl/src/run.ts [hunk 1]
```diff
@@ -228,7 +228,10 @@ async function main() {
       ...connectionInfo.driverOptions,
     });
   } catch (e: any) {
-    console.error(`${e?.name}: ${e?.message}`);
+    // for debugging
+    if (process.env.MONGOSH_DISPLAY_STARTUP_STACK_TRACE)
+      console.error(e?.stack);
+    else console.error(`${e?.name}: ${e?.message}`);
```
Comment (addaleax) [inline packages/cli-repl/src/run.ts:234]: This is just for us, it's pretty annoying to have the stack trace not showing up for errors occuring during startup

File: packages/e2e-tests/test/test-shell.ts [hunk 2]
```diff
@@ -160,7 +160,7 @@ export class TestShell {
     return this._rawOutput;
   }
 
-  get process(): ChildProcess {
+  get process(): ChildProcessWithoutNullStreams {
```
Comment (addaleax) [inline packages/e2e-tests/test/test-shell.ts:163]: Just a drive-by to fix type errors showing up in the e2e.spec.ts file

File: packages/cli-repl/src/mongosh-repl.ts [hunk 3]
```diff
@@ -658,8 +730,39 @@ class MongoshNodeRepl implements EvaluationListener {
     input: string,
     filename: string
   ): Promise<ShellResult> {
-    const { repl } = this.runtimeState();
-    return await promisify(repl.eval.bind(repl))(input, repl.context, filename);
+    const { repl, context } = this.runtimeState();
+    if (repl) {
+      return await promisify(repl.eval.bind(repl))(input, context, filename);
+    } else {
```
Comment (lerouxb) [inline packages/cli-repl/src/mongosh-repl.ts:736]: Suuuper nitpicky and I'm surprised that our linting didn't complain, but you can probably just remove the else and save one level of indentation because the if is guaranteed to return. But also just ignore me if you prefer this style.

File: packages/cli-repl/src/run.ts (hunk 1 above)
Comment (lerouxb) [reply packages/cli-repl/src/run.ts:234]: I should probably already know this answer, but why don't we display the stack trace by default all the time?

File: packages/cli-repl/src/mongosh-repl.ts (hunk 3 above)
Comment (addaleax) [reply packages/cli-repl/src/mongosh-repl.ts]: Yeah, thanks for pointing this out! Addressed this in 8f7f38f :)

**Review by lerouxb — APPROVED** [review]

File: packages/cli-repl/src/cli-repl.ts [hunk 4]
```diff
@@ -186,8 +186,17 @@ export class CliRepl implements MongoshIOProvider {
       this.bus.emit('mongosh:error', err, 'io');
     });
 
+    let jsContext = this.cliOptions.jsContext;
+    if (jsContext === 'auto' || !jsContext) {
+      jsContext = CliRepl.getFileAndEvalInfo(this.cliOptions)
```
Comment (kmruiz) [inline packages/cli-repl/src/cli-repl.ts:191]: Do we want (didn't see where) to add the jsContext to our telemetry as a dimension to the startup event?

Comment (addaleax) [reply packages/cli-repl/src/cli-repl.ts:191]: @kmruiz Thanks, added it! This is probably not going to be very different from `is_interactive` in practice, but keeping track of the number of people who run it with `js_context === 'repl' && is_interactive === false` might be valuable

**Review by kmruiz — APPROVED** [review]

**Comment by MarcusElevait** [comment]
Why is this not versioned as a breaking change? This breaks javascript file executions that uses native things like `require("fs")`

**Comment by tiagoliveira9** [comment]
Yes, this should be versioned as a breaking change, I have several scripts impacted with these changes

**Comment by tiagoliveira9** [comment]
And the helper doesn't show up the new `--jsContext` option

**Comment by addaleax** [comment]
@MarcusElevait @tiagoliveira9 Just for clarity, this was unintended breakage and we released mongosh 2.2.1 soon after to fix it: https://github.com/mongodb-js/mongosh/releases/tag/v2.2.1
